### ISSUE TITLE
Add ability to create path aliases for packages and scoped packages 

### ISSUE THREAD

**Issue by prowwid** [npm/npm#16464]
#### I'm opening this issue because:

There is a restriction on organization and package names that do not allow to use upper-case characters in names. So using TypeScript, if I want to use upper-case names for organization and packages to have imports look like 
```javascript
import {Connection, Command} from "@MyCompany/MyPackage/Database";
```
I would need to use crutches that would work for a single application but won't be a solution for public packages that would be distributed to users.

#### What's the feature?
I suggest adding "pathAlias"(or something similar - option name is not so important)  option that would allow setting path alias of scope and package in node_modules.
For example, if we would have following package.json:
```php
{
    "name": "@my-company/database-access",
    "pathAlias": "@MyCompany/DatabaseAccess"
}
```
nmp would generate @MyCompany directory in the node_modules directory for scope and create DatabaseAccess directory in the scope for the package.
I think It's the easiest way to have this feature without BC breaks.

#### What problem is the feature intended to solve?
The absence of ability to set path mapping with a custom scope and package name different from the ones set in a package name.

#### Is the absence of this feature blocking you or your team? If so, how?
Basically yes. I want to distribute TypeScript packages and use a uppercase notation for everything.

#### Is this feature similar to an existing feature in another tool?
Some sort of functionality I'm referring to is implemented in [module-alias ](https://www.npmjs.com/package/module-alias) package.
But If we would talk about other languages, then yes. Just look at ["composer.json"](https://github.com/laravel/framework/blob/5.4/composer.json) (PHP package manager configuration) of Laravel framework. The name of the package is "laravel/framework" but in code, you would use classes like in the following example:
```php
 use Illuminate\Database\Eloquent\Model; // Namespace does not match package name and use a uppercase notation

class Flight extends Model {
   // some code
}
```

Using composer you need to specify "autoload" option and set "namespace to path" mapping to achieve the result from the example above:

```json
 "autoload": {
        "psr-4": {
            "Illuminate\\": "src/Illuminate/" 
        }
    },
````

#### Is this a feature you're prepared to implement, with support from the npm CLI team?
It would take time for me to get familiar with npm codebase and won't happen in the nearest future but for such feature, I'm ready to invest my time.

**Comment by prowwid** [comment]
Hello, any thoughts o this issue?

**Comment by zkat** [comment]
Hi!

I looked over this feature request a couple of times wonder what it is we _could_ do about this, and I think this is a very tricky feature to implement that breaks a lot of the assumptions the installer makes, and plays with fire in others.

As far as I can tell, the reason composer's feature works is because they use autoload files that can, for example, map to [psr-4](http://www.php-fig.org/psr/psr-4/), an actual PHP language feature.

npm itself has no control whatsoever over the module loading algorithm, and even if you tried to convince the Node.js folks, that would probably fall very flat: there are hundreds of thousands of packages (and many many applications) that rely on what has been very solid, mature behavior. The module loading spec, of all things, is probably the hardest thing to change in the node project.

So, doing anything like autoload itself is out of the question. What other restrictions are there, and what could we do?

Well, the only thing npm could do here is to set up some arcane sym/hardlink mechanism that would somehow allow this stuff to be arranged into place. That seems like an incredible amount of complexity added to the installer considering the perceived value. Not to mention there's no way to get it right: this sort of thing isn't OS-portable, and we'd have to compensate for odd conflicts when generating these aliases at all levels. This is why the package name restriction is there: because it caused an immense number of problems to deal with case-insensitive and case-normalizing filesystems like NTFS and HFS+. Path aliases are no better here than lifting the restriction.

That said, there's ways to achieve something like what you're asking for entirely in userland, which I believe is the right solution here: You can create an alias package with a concise package name (still lowercase), and then re-export anything you need from there -- much like composer's `autoload` seems to work.

You do have some more alternatives here: if you know you'll be using case-normalizing filesystems (basically if you're using Windows or OSX), you can remove the dashes from your path, and it'll Just Work™

For example, this would be perfectly valid:
```js
// in package.json
{
  "dependencies": "@mycompany/mypackage"
}

// In your code
import {Connection, Command} from "@MyCompany/MyPackage/MyClass"
```

The last alternative I can suggest, since it seems like you're using a build step, is to just install or write a plugin for your build tool that will convert the code. There's already one that does essentially the inverse of what you're asking over at https://npm.im/case-sensitive-paths-webpack-plugin.

tbh, that is probably the best/safest route to take. And it would allow you full control over how this feature gets implemented.

The tl;dr here is that no, I don't think this is the sort of thing that can/should go into npm. As with many other things, though, if it's implemented in userland and gains popularity, it's something we would then consider, depending on the way the user package decided to resolve these conflicts. Solving this at the npm level, though, is probably the worst and trickiest layer to implement this in (vs userland and vs Node.js).

Have a nice day, and genuine thanks for the detailed, considered, well-written feature request. <3

**Comment by prowwid** [comment]
Hi @zkat,

Thank you for the detailed answer. 
I fully understand the downsides of suggested solution but still hoped you wound consider the idea of camel-cased package names. I would proceed with the plugin solution to solve the issue.

Thanks again.

With best regards,
Dmity
